File: uploader_widget/uploader.py
import ipywidgets as widgets
from traitlets import Unicode, Int, Bool, Any
from traitlets import observe
import base64
from math import ceil

@widgets.register
class FileUploader(widgets.DOMWidget):
    """An example widget."""
    _view_name = Unicode('UploaderView').tag(sync=True)
    _model_name = Unicode('UploaderModel').tag(sync=True)
    _view_module = Unicode('uploader_widget').tag(sync=True)
    _model_module = Unicode('uploader_widget').tag(sync=True)
    _view_module_version = Unicode('^0.0.1').tag(sync=True)
    _model_module_version = Unicode('^0.0.1').tag(sync=True)

    base64_data_blob = Unicode().tag(sync=True)
    file= Any().tag(sync=True)
    next_slice = Int().tag(sync=True)
    file_size = Int().tag(sync=True)
    file_name = Unicode().tag(sync=True)
    file_type = Unicode().tag(sync=True)
    next_blob = Int().tag(sync=True)
    base64_data = Unicode().tag(sync=True)
    label = Unicode("Upload a File").tag(sync=True)
    new_file = Bool().tag(sync=True)
    disabled = Bool().tag(sync=True)
    data = ''

    @observe("new_file")
    def clear_data(self, change):
        if self.new_file:
            self.data = ''
            self.new_file = False

    @observe("base64_data")
    def _update_label(self, change):
        self.label = self._label() or self.label

    @observe("base64_data_blob")
    def save__blob_and_request_next(self, change):
        bytes_string = base64.b64decode(self.base64_data_blob.split(',')[1])
        self.data += bytes_string.decode(encoding="utf-8", errors="ignore")
        if self.next_slice < self.file_size:
            self.next_blob = self.next_slice
            self.label = f'Uploading {self.file_name} - {ceil((self.next_slice/self.file_size)*100)}%'
        else:
            self.label = f'{self.file_name} uploaded'


    def _label(self):
        return "{} bytes uploaded {}".format(
            len(self.base64_data),
            self.base64_data.split(";")[0])

File: js/lib/uploader.js
var cs = require('./css/style.css');
var widgets = require('@jupyter-widgets/base');
var _ = require('lodash');

// Custom Model. Custom widgets models must at least provide default values
// for model attributes, including
//
//  - `_view_name`
//  - `_view_module`
//  - `_view_module_version`
//
//  - `_model_name`
//  - `_model_module`
//  - `_model_module_version`
//
//  when different from the base class.

// When serialiazing the entire widget state for embedding, only values that
// differ from the defaults will be specified.
var UploaderModel = widgets.DOMWidgetModel.extend({
    defaults: _.extend(widgets.DOMWidgetModel.prototype.defaults(), {
        _model_name : 'UploaderModel',
        _view_name : 'UploaderView',
        _model_module : 'uploader_widget',
        _view_module : 'uploader_widget',
        _model_module_version : '0.0.1',
        _view_module_version : '0.0.1',
    })
});

// Custom View. Renders the widget model.
var UploaderView = widgets.DOMWidgetView.extend({
    className: "uploader-widget",
    events: {
        dragenter: "onDragEnter",
        dragleave: "onDragLeave",
        drop: "onDrop",
        "change input": "onChangeInput"
    },

    render: function() {
        this.$label = $("<div/>", {"class": "uploader-label"})
            .appendTo(this.$el);

        this.$fileField = $("<input/>", {type: "file"})
            .appendTo(this.$el);

        this.model.on('change:next_blob', this.uploadSlice, this);
        this.model.on('change:disabled', this.toggleDisabled, this);
        this.update();
    },
    toggleDisabled: function(){
        if(this.model.get("disabled")==true) {
            this.$fileField[0].setAttribute("disabled", "true");
            this.$label[0].parentElement.style.opacity = "0.3";
        } else {
            this.$fileField[0].removeAttribute("disabled");
            this.$label[0].parentElement.style.opacity = "0.6";
        }

    },
    onChangeInput: function(){
        this.uploadFile(this.$fileField[0].files[0]);
    },
    setFile: function(file){
        var reader  = new FileReader();

        reader.addEventListener("load", _.bind(function(){
            this.model.set("base64_data", reader.result);
            this.touch();
        }, this), false);

        reader.readAsDataURL(file);
    },
    uploadFile: function(file){
        this.model.set("file", file);
        var file = this.model.get("file");
        var extension = file.name.match(/\.[0-9a-z]+$/i)[0];
        this.model.set('new_file', true);
        this.model.set('file_type', extension);
        this.model.set('file_name', file.name);
        this.model.set('file_size', file.size);
        this.model.set("next_blob", 0);
        this.uploadSlice();
    },
    uploadSlice: function(){
        var reader = new FileReader();
        var start = this.model.get('next_blob');
        var slice_size = 1000 * 1024;
        var next_slice = start + slice_size + 1;
        var file = this.model.get("file")
        var blob = file.slice(start, next_slice);

        reader.addEventListener("load", _.bind(function(){
            this.model.set("base64_data_blob", reader.result);
            this.touch();
        }, this), false);

        this.model.set('next_slice', next_slice);
        reader.readAsDataURL(blob);
    },
    update: function() {
        this.$label.html(this.model.get("label"));
    },
    onDragEnter: function(evt){ this.$el.addClass("uploader-dragged");},
    onDragLeave: function(evt){ this.$el.removeClass("uploader-dragged");},
    onDrop: function(evt){
        evt.preventDefault();
        evt.stopImmediatePropagation();
        this.$el.removeClass("uploader-dragged");
        this.uploadFile(evt.dataTransfer.files[0]);
    }
});

module.exports = {
    UploaderModel : UploaderModel,
    UploaderView : UploaderView
};
